<?xml version="1.0"?>

<!-- rover_gazebo_keyboard_ctrl.launch

Use Gazebo to simulate and control a rover with Ackermann steering
using keyboard commands.

Arguments:
  namespace (string, default: rover)
    rover namespace.
-->

<launch>
    <node name="odometry" pkg="rover_gazebo_control" type="odometry.py" output="screen" required="true"/>

    <arg name="namespace" default="rover"/>
    <arg name="cmd_timeout" default="0.5"/>
    <arg name="node_start_delay" default="5"/>
    <arg name="publish_frequency" default="30.0"/>
    <arg name="linear_velocity" default="10.0"/>
    <arg name="angular_velocity" default="1.3"/>
    <arg name="model" default="$(find rover_desc_pkg)/urdf/rover_robot.xacro"/>

    Launch the Gazebo simulation of the rover.
    <include file="$(find rover_gazebo_control)/launch/rover_gazebo_control.launch">
        <arg name="namespace" value="$(arg namespace)"/>
    </include>
    <rosparam file="$(find rover_desc_pkg)/config/joint.yaml"
              command="load"/>


    Control the steering and wheel joints.
    <node name="rover_controller" pkg="rover_gazebo_control" type="rover_controller.py"
          output="screen" ns="$(arg namespace)">
        <param name="cmd_timeout" value="$(arg cmd_timeout)"/>
        <param name="publish_frequency" value="$(arg publish_frequency)"/>
        <rosparam file="$(find rover_gazebo_control)/config/rover_ackermann_ctrlr_params.yaml" command="load"/>
    </node>

    <!-- Keyboard control node. -->
    <node name="rover_teleop" pkg="rover_gazebo_control" type="rover_teleop.py"
          output="screen" ns="$(arg namespace)">
        <param name="hz" value="$(arg publish_frequency)"/>
        <param name="steps" value="10"/>
        <param name="linear_velocity" value="$(arg linear_velocity)"/>
        <param name="angular_velocity" value="$(arg angular_velocity)"/>
        <param name="robot_description" command="$(find xacro)/xacro.py $(arg model)"/>
    </node>

    <!-- Node to convert /cmd_vel to /ackermann_command -->
    <node name="cmd_vel_to_ackermann_drive" pkg="rover_gazebo_control"
          type="cmd_vel_to_ackermann_drive.py" output="screen" ns="$(arg namespace)"
          launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@'"/>


</launch>
